todo_app/app.py

```diff
@@ -1,31 +1,120 @@
-from flask import Flask, render_template, request, redirect, url_for
+from functools import wraps
+from os import abort
+
+import requests
+
+from flask import Flask, render_template, request, redirect, url_for, abort
+
+from todo_app import flask_config
+from todo_app.User import User
 from todo_app.mongo_repository import MongoRepository
 from todo_app.viewmodel import ViewModel
+from flask_login import LoginManager, login_required, login_user, current_user
+from oauthlib.oauth2 import WebApplicationClient
 
 
 def create_app():
     app = Flask(__name__)
     app.config.from_object('todo_app.flask_config.Config')
+    config = flask_config.Config()
     mongo_repo = MongoRepository()
 
+    login_manager = LoginManager()
+    app.config['SECRET_KEY'] = config.SECRET_KEY
+    login_manager.init_app(app)
+    login_manager.login_view = "login"
+    client = WebApplicationClient(config.GITHUB_OAUTH_CLIENT_ID)
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(client.prepare_request_uri('https://github.com/login/oauth/authorize'))
+
+    # Github OAuth flow when unauthenticated
+    @login_manager.user_loader
+    def load_user(user_id):
+        return mongo_repo.get_user(user_id)
+
+    def authorise_user(role):
+        def authorized(f):
+            @wraps(f)
+            def wrapper(*args, **kwargs):
+                if hasattr(current_user, 'role') and current_user.role != role:
+                    abort(401)
```

Comment:
Usually 401 Unauthorized means the user hasn't authenticated at all. 403 Forbidden would fit better here - they have logged in, but they don't have permission.